### PR TITLE
fix(msteams): Don't schedule important events

### PR DESCRIPTION
**Pull request by ykamo001**
MSTeams sends important events that we need to process right away. Don't schedule them for processing since it can take too long and expire the event, or even the JWT. Added logic to help let a subsection of them go through to the synchronous handler.

### CODE REVIEW

File: src/sentry/integrations/msteams/webhook.py [hunk 1]
```diff
@@ -238,6 +238,25 @@ def __init__(self, **kwargs) -> None:
     def dispatch(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
         return super().dispatch(request, *args, **kwargs)
 
+    def post(self, request: HttpRequest) -> HttpResponse:
```
Comment (ykamo001) [inline src/sentry/integrations/msteams/webhook.py:241]: Just moving this function to the top so it's easier to read

File: src/sentry/middleware/integrations/parsers/msteams.py [hunk 2]
```diff
@@ -95,6 +111,9 @@ def get_response(self) -> HttpResponseBase:
             logger.info("%s.no_regions", self.provider, extra={"path": self.request.path})
             return self.get_response_from_control_silo()
 
+        if self._check_if_event_should_be_sync(data=self.request_data):
+            return self.get_default_missing_integration_response()
```
Comment (GabeVillalobos) [inline src/sentry/middleware/integrations/parsers/msteams.py:115]: Interesting, so we don't handle synchronous events on our end at all? I'm also curious if MSTeams does anything if too many 4xx or 5xx errors are surfaced. Some other integrations like gitlab will eventually disable the webhook integration automatically if this occurs.

**Review by GabeVillalobos — APPROVED** [review]
One small question, but otherwise the changes look good to me

File: src/sentry/middleware/integrations/parsers/msteams.py (hunk 2 above)
Comment (ykamo001) [reply src/sentry/middleware/integrations/parsers/msteams.py:115]: We handle them synchronously if the integration cannot be inferred or if the organization does not have regions. Basically anything that calls `get_response_from_control_silo` is the synchronous method

Comment (ykamo001) [reply src/sentry/middleware/integrations/parsers/msteams.py:115]: Just noticed the function i'm calling is wrong for the response lol

**Review by dashed — APPROVED** [review]

**Comment by sentaur-athena** [comment]
@ykamo001 if the event takes so long to happen async shouldn't we send other ones sync too?

**Comment by ykamo001** [comment]
> @ykamo001 if the event takes so long to happen async shouldn't we send other ones sync too?

There are actions, like resolve an issue, that need to happen async and cannot happen in the control silo. We will likely have to fine tune this as we learn about the integration

File: src/sentry/middleware/integrations/parsers/msteams.py [hunk 3]
```diff
@@ -46,17 +53,29 @@ def get_integration_from_request(self) -> Integration | None:
...
+    def _check_if_event_should_be_sync(cls, data: Mapping[str, Any]) -> bool:
+        """
+        Determine if an event should be handled synchronously, or if we can defer to async.
+        """
+        raw_event_type = data.get("type", None)
+        if raw_event_type is None:
+            return True
+
+        event_type = MsTeamsEvents.get_from_value(value=raw_event_type)
+        return event_type in cls._synchronous_events
```
Comment (iamrajjoshi) [inline src/sentry/middleware/integrations/parsers/msteams.py:66]: just a question - why is installation the only event required to be handled synchronously?

Comment (ykamo001) [reply src/sentry/middleware/integrations/parsers/msteams.py:66]: This is the event we get during an installation for teams, and we want to unblock them first so they can properly install and setup issue alerts

**Pull request merged**